p4-16/spec/P4-16-spec.mdk

```diff
@@ -1460,6 +1460,123 @@ lvalue
 The following is a legal l-value: `headers.stack[4].field`.  Note
 that method and function calls cannot return l-values.
 
+## Named parameters {#sec-named-params}
+
+P4 supports named parameters to enable invoking constructors, controls,
+parsers, and externs with a subset of the arguments when appropriate default
+values are provided. Specifically:
+
+- Calls specify parameters by name. By default, all parameters must be
+  specified.
+- Optional parameters are specified by assigning a default parameter value:\
+  `param_modifier param_type name = default_parameter_value`
+- Mandatory parameters must preceed all optional parameters and all parameters
+  must be called in the order in which they were declared. This is similar to
+  C++, except that preceeding optional parameters can be fully omitted (as in
+  Python).
+- We only allow optional parameters for data-plane program entities that are
+  not exposed at runtime, i.e., do not participate in the generation of
+  control-plane APIs. These include:
+   - Constructors, including extern constructors and package constructors,
+     controls, and parsers
+   - Extern methods
+   - Internal actions or functions that are used exclusively in the data-plane
+- Optional parameters must be either `in` or directionless
+
+### Default values
+
+If `_` is provided as the default value the compiler will initialize the
+optional parameter with an appropriate value.
+
+In addition, we define the value `None`as a keyword to specify a parameter for
```

Comment:
Adding `None` as a keyword is going to break any program that already uses `None` as a name anywhere.

Reply:
Yes, that could be a problem. However, I have seen no uses on None in the entire body of p4_16 code that I have access to.

Reply:
I really don't think we need to have a None default value. We don't need it for anything.
It is much simpler to just an annotation of `@optional` on parameters that do not need to have a value specified, as it is today.  However, we can only allow such parameters for objects that are interpreted by the back-end, e.g., package or extern constructor parameters.
We can have default parameter values, but they have to have the right type.

Reply:
It is not clear to me why would you prefer an annotation over explicitly stating the optionality in the language. Just because we currently have an experimental feature implemented in the compiler? I can promise that we implement the new feature instead.

Line 1477 states that all optional parameters are applicable to objects not exposed to the control plane. If it is not coming across, please suggest different wording.